src/pe/authenticode.rs

```diff
@@ -76,24 +96,155 @@ impl<'s> Iterator for ExcludedSectionsIter<'s> {
+                        // 5. Hash everything from the end of the checksum field to immediately before the start
+                        //    of the Certificate Table entry, as specified in Optional Header Data Directories.
+                        let out =
+                            Some(&bytes[checksum_end..sections.datadir_entry_certtable.start]);
+                        debug!(
+                            "hashing {checksum_end:#x} {:#x}",
+                            sections.datadir_entry_certtable.start
+                        );
+
+                        // 6. Get the Attribute Certificate Table address and size from the Certificate Table entry.
+                        //    For details, see section 5.7 of the PE/COFF specification.
+                        // 7. Exclude the Certificate Table entry from the calculation
+                        self.state =
+                            IterState::CertificateTableEnd(sections.datadir_entry_certtable.end);
+
+                        return out;
+                    }
+                    IterState::CertificateTableEnd(start) => {
+                        // 7. Exclude the Certificate Table entry from the calculation and hash everything from
+                        //    the end of the Certificate Table entry to the end of image header, including
+                        //    Section Table (headers). The Certificate Table entry is 8 bytes long, as specified
+                        //    in Optional Header Data Directories.
+                        let end_image_header = sections.end_image_header;
+                        let buf = Some(&bytes[start..end_image_header]);
+                        debug!("hashing {start:#x} {:#x}", end_image_header - start);
+
+                        // 8. Create a counter called SUM_OF_BYTES_HASHED, which is not part of the signature.
+                        //    Set this counter to the SizeOfHeaders field, as specified in
+                        //    Optional Header Windows-Specific Field.
+                        let sum_of_bytes_hashed = end_image_header;
+
+                        self.state = IterState::HeaderEnd {
+                            end_image_header,
+                            sum_of_bytes_hashed,
+                        };
+
+                        return buf;
+                    }
+                    IterState::HeaderEnd {
+                        end_image_header,
+                        sum_of_bytes_hashed,
+                    } => {
+                        // 9. Build a temporary table of pointers to all of the section headers in the
+                        //    image. The NumberOfSections field of COFF File Header indicates how big
+                        //    the table should be. Do not include any section headers in the table whose
+                        //    SizeOfRawData field is zero.
+                        let mut sections: VecDeque<SectionTable> = self
```

Comment:
i'd not expect allocations in the iterator itself; (same comment as above); is there anyway to avoid allocating? (if there's not, or it's too difficult, that's fine, just something to maybe document?)